Fixes to ionic strength, pH scale conversion, fugfac, Ft etc.

jamesorr opened the pull request:
Jean-Pierre,

I made numerous modifications to seacarb v3.0.3 so that in our comparison seacarb results would agree with other critical packages (e.g., CO2SYS) in the comparison.  One of these changes (replacing use of incorrect ionic strength 'ION' with correct version 'iom0' in Kf.R) did not seem to get carried over from what I sent you many months ago. Perhaps it was forgotten because we were not yet following standard github practice with pull requests etc.

More generally, errors in carb.R, K1.R, and K2.R caused significant discrepancies. Those errors probably originated with major changes made for seacarb 3.0.   It took some time to find and correct the problems, but now results are right on target: they agree with CO2SYS, mocsy, etc. Here are a few general details about the major changes:

1) theta.R - new routine to compute potential temperature (relies on 'swTheta' from 'oce' package);
2) rho.R - convert input temperature T (ITS 90) to T (IPTS 68) & then compute density, as intended;
3) global change - use more accurate Ft equation and less ambiguous equation for St (at.R, Ks.R, Pcorrect.R, kconv.R, pHconv.R, buffer.R, buffesm.R, carb.R, ...);
4) Kb. R - refined pH scale conversion;
5) K1.R, K2.R - corrected errors in pH scale conversion; and
6) carb.R - corrected errors in pH scale conversion & fCO2 -> pCO2 conversion

jpgattuso commented:
MANY thanks for your hard work with seacarb. I have merged your additions and made other changes to the man files, the description file (adding new dependency) and changelog.

I think that there is an error in the code of carb.R where there is:
      # In situ pCO2 - compute with potential T & surface P (atm)  
      # Potential pCO2 - compute with potential T & surface P  

It should read as follows, right ? (I have made the change)
      # In situ pCO2 - compute with in situ T & total in situ P
      # Potential pCO2 - compute with potential T & surface P (atm)

I have produce version 3.0.4. Could you check the ChangeLog (also pasted below) before I submit to CRAN?

-----------------------------------------------------------------------------
version 3.0.4, 2015-01-04
-----------------------------------
- Function carb: added potential fCO2 and pCO2 (fCO2pot, pCO2pot) in the output
- Function carb: added a new option "gas" as an input argument as recommended by Orr & Epitalon (2015). It can be either 'insitu' or 'potential' (the default). This new option describes whether in situ or potential fCO2 and pCO2 should be computed. In situ values are calculated after making pressure corrections to K0 and the fugacity coefficient following Weiss (1974) using total in situ pressure and in situ temperature. Potential values are calculated at 1 atm pressure while using potential instead of in situ temperature.
- Function carb: as a result of the change above, the carb function now uses the swTheta function of the oce package which therefore needs to be installed.
- Functions carb, buffer, buffesm and Ks: removed unused variables ION and iom0
- Function Kf: fixed erroneous equation using wrong ionic strength ION
- Changed data set Pcoeffs.txt: changed 'a0' for Kspa from -46.00 to -45.96, consitent with literature
- New function: tetha to compute potential temperature (corresponding help file added too)
- Function rho: convert input T (ITS 90) to T (IPTS 68) & then compute density, as intended
- Functions Ks, Pcorrect, at, buffer, buffesm, kconv, pHconv: more accurate Ft equation and clearer St equation (best practices)
- Function Kb: refined pH scale conversion
- Changed K1, K2, carb: corrected errors in pH scale conversion
- Corrected English
- Function carb: corrected errors in pH scale conversion & fCO2 -> pCO2 conversion

jamesorr commented:
Bonjour Jean-Pierre,

Very sorry to hear that you were ill over the holidays!  Sounds like a
nasty case of bronchitis. I hope that you have almost recovered!

It would be great if you could make seacarb 3.0.4 available by tomorow, as
that will make it much easier for me to generate seacarb's computed
variables from the GLODAP gridded data (using exisitng scripts). Below are
my suggested changes for your ChangeLog message.

   - Function carb: added potential fCO2 and pCO2 (fCO2pot, pCO2pot) in the
   output
   - Function carb: added a new option "gas" as an input argument as
   recommended by Orr & Epitalon (2015). It can be either 'insitu' or
   'potential' (the default). For flags 21 to 25, this new option describes
   whether the input pCO2 that is given is "in situ" pCO2 or potential pCO2.
   For other flags, this input option is not used. In situ fCO2 and pCO2 are
   calculated after making pressure corrections to K0 and the fugacity
   coefficient following Weiss (1974) using total in situ pressure and in situ
   temperature. Potential fCO2 and pCO2 are also calculated but at 1 atm
   pressure while using potential temperature instead of in situ temperature.
   - Function carb: as a result of the change above, the carb function now
   calls a new function 'theta', which itself calls the 'swTheta' function
   from the 'oce' package. Hence that package also needs to be installed.
   - Functions carb, buffer, buffesm and Ks: removed unused variables ION
   and iom0
   - Function Kf: fixed erroneous equation using wrong ionic strength
   (replaced ION with iom0)
   - Changed data set Pcoeffs.txt: changed 'a0' for Kspa from -46.00 to
   -45.96, consitent with the literature
   - New function: 'theta' to compute potential temperature (corresponding
   help file added too)
   - Function rho: convert input T (ITS 90) to T (IPTS 68) & then compute
   density, as intended
   - Functions Ks, Pcorrect, at, buffer, buffesm, kconv, pHconv: more
   accurate Ft equation and clearer St equation (best practices)
   - Function Kb: refined pH scale conversion
   - Changed K1, K2, carb: corrected errors in pH scale conversion
   - Corrected English
   - Function carb: corrected errors in pH scale conversion & fCO2 -> pCO2
   conversion

Amitiés,

Jim

On Sun, Jan 4, 2015 at 8:25 AM, Jean-Pierre Gattuso <
notifications@github.com> wrote:

> MANY thanks for your hard work with seacarb. I have merged your additions
> and made other changes to the man files, the description file (adding new
> dependency) and changelog.
>
> I think that there is an error in the code of carb.R where there is:
> # In situ pCO2 - compute with potential T & surface P (atm)
>
> # Potential pCO2 - compute with potential T & surface P
>
> It should read as follows, right ? (I have made the change)
> # In situ pCO2 - compute with in situ T & total in situ P
> # Potential pCO2 - compute with potential T & surface P (atm)
>
> I have produce version 3.0.4. Could you check the ChangeLog (also pasted
> below) before I submit to CRAN?
> ------------------------------
> version 3.0.4, 2015-01-04
>
>    - Function carb: added potential fCO2 and pCO2 (fCO2pot, pCO2pot) in
>    the output
>    - Function carb: added a new option "gas" as an input argument as
>    recommended by Orr & Epitalon (2015). It can be either 'insitu' or
>    'potential' (the default). This new option describes whether in situ or
>    potential fCO2 and pCO2 should be computed. In situ values are calculated
>    after making pressure corrections to K0 and the fugacity coefficient
>    following Weiss (1974) using total in situ pressure and in situ
>    temperature. Potential values are calculated at 1 atm pressure while using
>    potential instead of in situ temperature.
>    - Function carb: as a result of the change above, the carb function
>    now uses the swTheta function of the oce package which therefore needs to
>    be installed.
>    - Functions carb, buffer, buffesm and Ks: removed unused variables ION
>    and iom0
>    - Function Kf: fixed erroneous equation using wrong ionic strength ION
>    - Changed data set Pcoeffs.txt: changed 'a0' for Kspa from -46.00 to
>    -45.96, consitent with literature
>    - New function: tetha to compute potential temperature (corresponding
>    help file added too)
>    - Function rho: convert input T (ITS 90) to T (IPTS 68) & then compute
>    density, as intended
>    - Functions Ks, Pcorrect, at, buffer, buffesm, kconv, pHconv: more
>    accurate Ft equation and clearer St equation (best practices)
>    - Function Kb: refined pH scale conversion
>    - Changed K1, K2, carb: corrected errors in pH scale conversion
>    - Corrected English
>    - Function carb: corrected errors in pH scale conversion & fCO2 ->
>    pCO2 conversion
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jpgattuso/seacarb-git/pull/3#issuecomment-68624011>.
>
